### PR TITLE
Add Vec::leak

### PR DESCRIPTION
**Pull request by cramertj**

### CODE REVIEW

File: src/liballoc/vec.rs [hunk 1]
```diff
@@ -1367,6 +1367,43 @@ impl<T> Vec<T> {
...
+    /// leak.
+    ///
+    /// # Examples
+    ///
+    /// Simple usage:
+    ///
+    /// ```
+    /// fn main() {
+    ///     let x = vec![1, 2, 3];
+    ///     let static_ref: &'static mut [usize] = Box::leak(x);
```
Comment (jonas-schievink) [inline src/liballoc/vec.rs]:
```suggestion
    ///     let static_ref: &'static mut [usize] = Vec::leak(x);
```

**Comment by withoutboats** [comment]
@bors r+

Seems reasonable at least on nightly, its annoying to write by hand.

**Comment by matklad** [comment]
Am I missing something, or is this just `Box::leak(vec.into_boxed_slice())`?

**Comment by Centril** [comment]
@bors r-

This can be written without the `unsafe` as per @matklad's comment (even `Box::leak(vec.into())` will suffice) and so we should do that in the implementation to reduce the overall trusted computing base.

**Comment by cramertj** [comment]
> This can be written without the unsafe

Done, thanks for the suggestion.

**Comment by Centril** [comment]
Thanks!

@bors r=withoutboats,Centril rollup

File: src/liballoc/vec.rs [hunk 2]
```diff
@@ -1367,6 +1367,38 @@ impl<T> Vec<T> {
...
+    ///
+    /// This function is mainly useful for data that lives for the remainder of
+    /// the program's life. Dropping the returned reference will cause a memory
+    /// leak.
+    ///
+    /// # Examples
+    ///
+    /// Simple usage:
+    ///
+    /// ```
```
Comment (Centril) [inline src/liballoc/vec.rs]:
```suggestion
    /// ```
    /// #![feature(vec_leak)]
    ///
```

**Comment by Centril** [comment]
@bors r-

Failed in https://github.com/rust-lang/rust/pull/62216.

File: src/liballoc/vec.rs [hunk 3]
```diff
@@ -1367,6 +1367,40 @@ impl<T> Vec<T> {
...
+    /// fn main() {
+    ///     let x = vec![1, 2, 3];
+    ///     let static_ref: &'static mut [usize] = Vec::leak(x);
+    ///     static_ref[0] += 1;
+    ///     assert_eq!(static_ref, &[2, 2, 3]);
+    /// }
+    /// ```
+    #[unstable(feature = "vec_leak", issue = "62195")]
+    #[inline]
+    pub fn leak<'a>(vec: Vec<T>) -> &'a mut [T]
```
Comment (czipperz) [inline src/liballoc/vec.rs]: Why aren't we using `'static` here?

Comment (cramertj) [reply src/liballoc/vec.rs]: same reason as `Box::leak`, and mentioned in the comment (not WF if `T` doesn't outlive `'static`).

Comment (Lucretiel) [reply src/liballoc/vec.rs]: If the use case for this method is "for data that lives for the remainder of the program", is there ever a practical case where someone would want to leak a non-static T?

Comment (czipperz) [reply src/liballoc/vec.rs]: Well I'd assume someone might want to "leak" the memory and then put it into a custom data structure (ie a `Box`).

**Pull request merged**